Update dietpi-imager: add check free disk space

sal666 opened the pull request:
Try to stimate the space needed to store the image and the 7z file

sal666 committed:
Update dietpi-imager: add check of free disk space

sal666 committed:
Update dietpi-imager

MichaIng commented on .meta/dietpi-imager:
```diff
@@ -261,6 +261,35 @@ Do you want to overwrite or backup the existing file to /root/$OUTPUT_IMG_NAME.$
+
+		fi
+		NEEDED_FREE_SPACE=$(( $NEEDED_FREE_SPACE / 1024 / 1024 ))
+		CURRENT_FREE_SPACE=$(( $(df --output=avail /root | tail -n1 | mawk '{print $1}') / 1024 ))
```
@sal666 
Lets use `G_CHECK_FREESPACE`: https://github.com/MichaIng/DietPi/blob/dev/dietpi/func/dietpi-globals#L1807
```
G_CHECK_FREESPACE /root $NEEDED_FREE_SPACE || Exit_On_Fail
```
```
2019-11-21 21:43:02 root@micha:/var/lib# G_CHECK_FREESPACE /root 1000
[  OK  ] Free space check: path=/root | available=58753 MiB | required=1000 MiB
2019-11-21 21:43:16 root@micha:/var/lib# G_CHECK_FREESPACE /root 100000
[FAILED] Free space check: path=/root | available=58753 MiB | required=100000 MiB
```

The method is is exactly the same, output same information 😃.

sal666 replied:
Nice

MichaIng commented on .meta/dietpi-imager:
```diff
@@ -261,6 +261,35 @@ Do you want to overwrite or backup the existing file to /root/$OUTPUT_IMG_NAME.$
+		# Image file source
+		if [[ $FP_SOURCE_IMG ]]; then
+
+			NEEDED_FREE_SPACE=$(( $IMAGE_SIZE / 100 * 15 ))
```
@sal666 
You assume 7z compression ratio 15:100 right? Sounds realistic, however what you think about addition +100 MiB fixed (in every case) so we avoid to fill the partition to its limit? Could be done, with other suggestion:
```
G_CHECK_FREESPACE /root $(( $NEEDED_FREE_SPACE + 100 )) || Exit_On_Fail # Assure +100 MiB left partition space
```

sal666 replied:
Yep, sure

MichaIng commented on .meta/dietpi-imager:
```diff
@@ -261,6 +261,35 @@ Do you want to overwrite or backup the existing file to /root/$OUTPUT_IMG_NAME.$
+		# Drive source, clone with Clonezilla
+		else
+
+			NEEDED_FREE_SPACE=$(( $IMAGE_SIZE / 100 * 23 * 2 + 330000000 * 3 ))
```
@sal666 
Just as I am interested to understand the values:
23:100 compression of raw file space by Clonezilla import? This a second time within the final iso + ~315 MiB Clonezilla itself, once zip download, once extracted, once the final iso, roughly?

sal666 replied:
Exactly. Compression ratios are approximations by just one test. They could be adjusted to gain some extra margin. Something like 15->17 and 23->26.

MichaIng replied:
@sal666 
Since the contained file composition will only change marginally, it can be expected to stay very similar. If you rounded up your end result, it should be fine, especially when adding +100 MiB fixed.
_______________
For the Clonezilla size, it would be perfect if we could get the actual size from download URL:
```
2019-11-21 23:17:53 root@micha:/var/lib# CLONEZILLA_REPO='https://sourceforge.net/projects/clonezilla/files/clonezilla_live_alternative'
2019-11-21 23:18:02 root@micha:/var/lib# CLONEZILLA_VERSION=$(curl -s "$CLONEZILLA_REPO/" | grep 'class="folder "' | head -n1 | cut -d\" -f2)
2019-11-21 23:19:01 root@micha:/var/lib# CLONEZILLA_ZIP="clonezilla-live-${CLONEZILLA_VERSION}-amd64.zip"
2019-11-21 23:19:19 root@micha:/var/lib# curl -IL "$CLONEZILLA_REPO/${CLONEZILLA_VERSION}/${CLONEZILLA_ZIP}/download"
HTTP/2 302
server: nginx/1.14.0 (Ubuntu)
date: Thu, 21 Nov 2019 22:19:36 GMT
content-type: text/html; charset=UTF-8
location: https://downloads.sourceforge.net/project/clonezilla/clonezilla_live_alternative/20191024-eoan/clonezilla-live-20191024-eoan-amd64.zip?r=&ts=1574374776&use_mirror=netcologne
pragma: no-cache
cache-control: no-cache
x-ua-compatible: IE=edge,chrome=1
x-frame-options: SAMEORIGIN
content-security-policy: upgrade-insecure-requests
set-cookie: VISITOR=084d09e2-b1fc-4ce9-ae4e-eb83350b198a; Max-Age=315360000; Path=/; expires=Sun, 18-Nov-2029 22:19:36 GMT; HttpOnly
x-content-type-options: nosniff
strict-transport-security: max-age=31536000

HTTP/2 302
server: nginx/1.14.0 (Ubuntu)
date: Thu, 21 Nov 2019 22:19:37 GMT
content-type: text/html; charset=UTF-8
location: https://netcologne.dl.sourceforge.net/project/clonezilla/clonezilla_live_alternative/20191024-eoan/clonezilla-live-20191024-eoan-amd64.zip
content-disposition: attachment; filename="clonezilla-live-20191024-eoan-amd64.zip"
set-cookie: sf_mirror_attempt=clonezilla:netcologne:clonezilla_live_alternative/20191024-eoan/clonezilla-live-20191024-eoan-amd64.zip; Max-Age=120; Path=/; expires=Thu, 21-Nov-2019 22:21:37 GMT

HTTP/2 200
server: nginx/1.17.5
date: Thu, 21 Nov 2019 22:19:37 GMT
content-type: application/octet-stream
content-length: 314091868
last-modified: Thu, 24 Oct 2019 12:22:09 GMT
etag: "5db19771-12b8a95c"
accept-ranges: bytes

2019-11-21 23:21:24 root@micha:/var/lib# curl -sIL "$CLONEZILLA_REPO/${CLONEZILLA_VERSION}/${CLONEZILLA_ZIP}/download" | mawk '/^content-length:/{print $2}'
314091868
```
- So the related variables + final download URL need to be estimated prior to free space check for this.

sal666 committed:
Update dietpi-imager

WIP, needs some debugging

MichaIng committed:
DietPi-Imager | Fine tuning

+ When creating image from existing img file, check for sufficient disk space after truncating the file, since this can free more space than required by the final 7z archive easily.
+ When creating a Clonezilla-based installer image, now that the downloaded zip archive is removed immediately after being extracted, the are max two instances of Clonezilla stored at the disk at the same time: zip + tmpiso; tmpiso + iso; iso + 7z
  Hence required disk space can be reduced to: (( Clonezilla + DietPi ) * 2 ))
+ Within bash conditional double braces [[ ]], double quotes are not required, as white spaces are handled literally already. Leave double quotes around command arguments only (which is the reason why bourne shell single quotes require quoting, as "[" is a regular command, no special syntax).

MichaIng commented:
@sal666 
Is it ready to merge from your side?

sal666 commented:
Sure, go ahead!
